pkg/virt-handler/heartbeat/heartbeat.go

```diff
@@ -0,0 +1,163 @@
+package heartbeat
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sync"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	utilwait "k8s.io/apimachinery/pkg/util/wait"
+	k8scli "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	v1 "kubevirt.io/client-go/api/v1"
+	"kubevirt.io/client-go/log"
+	virtutil "kubevirt.io/kubevirt/pkg/util"
+	virtconfig "kubevirt.io/kubevirt/pkg/virt-config"
+	device_manager "kubevirt.io/kubevirt/pkg/virt-handler/device-manager"
+)
+
+type HeartBeat struct {
+	clientset                 k8scli.CoreV1Interface
+	deviceManagerController   device_manager.DeviceControllerInterface
+	clusterConfig             *virtconfig.ClusterConfig
+	host                      string
+	cpuManagerPaths           []string
+	lock                      *sync.Mutex
+	devicePluginPollIntervall time.Duration
+	devicePluginWaitTimeout   time.Duration
+}
+
+func NewHeartBeat(clientset k8scli.CoreV1Interface, deviceManager device_manager.DeviceControllerInterface, clusterConfig *virtconfig.ClusterConfig, host string) *HeartBeat {
+	return &HeartBeat{
+		clientset:               clientset,
+		deviceManagerController: deviceManager,
+		clusterConfig:           clusterConfig,
+		host:                    host,
+		// This is a temporary workaround until k8s bug #66525 is resolved
+		cpuManagerPaths:           []string{virtutil.CPUManagerPath, virtutil.CPUManagerOS3Path},
+		lock:                      &sync.Mutex{},
+		devicePluginPollIntervall: 1 * time.Second,
+		devicePluginWaitTimeout:   10 * time.Second,
+	}
+}
+
+func (h *HeartBeat) Run(heartBeatInterval time.Duration, stopCh chan struct{}) (done chan struct{}) {
+	done = make(chan struct{})
+	go func() {
+		h.heartBeat(heartBeatInterval, stopCh)
+		close(done)
+	}()
+	return done
+}
+
+func (h *HeartBeat) heartBeat(heartBeatInterval time.Duration, stopCh chan struct{}) {
+	// ensure that the node is synchronized with the actual state
+	// especially setting the node to unschedulable if device plugins are not yet ready is very important
+	// otherwise workloads get scheduled but are immediately terminated by the kubelet
+	h.do()
+	// Now wait for 10 seconds for the device plugins  to be initialized
+	// This is more than fast enough to be not treated as unschedulable by the cluster
+	// and ensures that the cluster gets marked as scheduled as soon as the device plugin is ready
+	if !h.waitForDevicePlugins(stopCh) {
+		return
+	}
```

Comment:
so since we're calling "return" here, if this fails, then the heartbeat silently never starts and won't recover?

Reply:
there is still something wrong like the test results show, but not sure if this is the issue: `false` should  ever only return in the case of the stopChan being closed.

Reply:
rebased and retesting. Looking at the failures I am not sure if something is wrong.

Reply:
> false should ever only return in the case of the stopChan being closed.

i'm sure that's the case, it's concerning looking though. It appears like there's a path that will result in the heartbeat never starting, which would result in the node never becoming schedulable.  Practically, that may not be possible, but the code as it is here doesn't instill confidence.

is there a way to ensure we eventually heartbeat even if device plugins don't initialize immediately for some reason?

Reply:
I can try to structure things differently, but 

> is there a way to ensure we eventually heartbeat even if device plugins don't initialize immediately for some reason?

is exactly what is happening. So do you think that the code needs device plugins started to start the regular heartbeat, or do you just think that it is not clear enough that this is not the case?